### PR TITLE
[release/3.1] Generate non-stable version for installer

### PR DESCRIPTION
**Pull request by wtgodbe**
Generate a non-stable ProductVersion for installers so that updates can occur between installers in the same Major.Minor.Patch band (e.g. previewN to previewN+1)

https://dev.azure.com/dnceng/internal/_build/results?buildId=821089&view=results

### CODE REVIEW

File: src/Installers/Windows/Wix.targets [hunk 1]
```diff
@@ -2,7 +2,7 @@
   <!-- Set versioning properties after Arcade SDK targets have been imported. -->
   <PropertyGroup>
     <!-- Used for generating stable upgrade codes for bundles -->
-    <Version>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).$(AspNetCorePatchVersion).0</Version>
+    <Version>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).$(AspNetCorePatchVersion).$(FileRevisionVersion)</Version>
```
Comment (joeloff) [inline src/Installers/Windows/Wix.targets:5]: I don't think this is what's used. For 3.1, we should fix this:
```
<GenerateGuid NamespaceGuid="$(NamespaceGuid)" Values="$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion);$(Platform)">
  <Output TaskParameter="Guid" PropertyName="BundleUpgradeCode" />
</GenerateGuid>
```

For example, if we're building the following bundles:

3.1.0.0 x64 should yield GUID A
3.1.20.78567 x64 should yield GUID A
3.1.20.78567 x86 should yield GUID B (x64 is SxS with x86) within the same major.minor range)
3.2.0.0 x64 should yield GUID C (making it SxS with any 3.1 build)

Comment (dougbu) [reply src/Installers/Windows/Wix.targets:5]: Just curious: What sets `$(FileRevisionVersion)`❔

And, a nit if we need this change:
```suggestion
    <Version>$(AspNetCoreMajorMinorVersion).$(AspNetCorePatchVersion).$(FileRevisionVersion)</Version>
```

Comment (joeloff) [reply src/Installers/Windows/Wix.targets:5]: https://github.com/dotnet/aspnetcore/blob/0250c22b9a8d7739ce4e2c75dd3e9d4c74e74f80/src/Installers/Windows/Wix.targets#L7

**Comment by wtgodbe** [comment]
Build of this branch: https://dev.azure.com/dnceng/internal/_build/results?buildId=821646&view=results

Build of this branch with patch version +1: https://dev.azure.com/dnceng/internal/_build/results?buildId=821647&view=results

**Review by dougbu — COMMENTED** [review]
Again, WFM but @joeloff's opinion is the one that matters

**Comment by wtgodbe** [comment]
New builds w/ version+1 (last one had a packageIndex failure)

https://dev.azure.com/dnceng/internal/_build/results?buildId=823563&view=results
https://dev.azure.com/dnceng/internal/_build/results?buildId=823562&view=results

**Comment by wtgodbe** [comment]
@joeloff I pushed https://github.com/dotnet/aspnetcore/pull/26065/commits/1f7fb7c23518b5b6106c3962b293e37ccce1aaaf. This will go into 3.1.10, so we'll need to add the upgrade codes for 3.1.9 once those are available.

**Review by joeloff — APPROVED** [review]
I think this looks good. We should get a couple of test builds to look at

Maj.Min.x
Maj.Min.x+1
Maj.Min+1

and pick two random official releases and then check that downgrades are blocked, upgrades are processed.

**Comment by wtgodbe** [comment]
Builds:
3.1.9: https://dev.azure.com/dnceng/internal/_build/results?buildId=823650&view=results, https://dev.azure.com/dnceng/internal/_build/results?buildId=823651&view=results

3.1.10: https://dev.azure.com/dnceng/internal/_build/results?buildId=823651&view=results


3.2.0: https://dev.azure.com/dnceng/internal/_build/results?buildId=823660&view=results

File: src/Installers/Windows/SharedFrameworkBundle/Bundle.wxs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (dougbu) [inline src/Installers/Windows/SharedFrameworkBundle/Bundle.wxs:2]: nit: Ick❕

**Review by dougbu — APPROVED** [review]
If manual tests of upgrades work, WFM

**Comment by wtgodbe** [comment]
Testing confirmed that the behavior is now correct. 3.1.10 removes 3.1.9 when installed, 3.1.9 won't install when 3.1.10 is present, and 3.2.0 installs alongside 3.1.10.

**Comment by joeloff** [comment]
Woohoo! That's great to hear @wtgodbe

**Review by joeloff — APPROVED** [review]

**Comment by dougbu** [comment]
@Pilchie this is the PR where I meant to ask you about 3.1.10 ask or tell mode

**Comment by Pilchie** [comment]
Same - tell mode with an explicit mail should be fine.

**Pull request merged**